internal/pkg/policy/monitor.go

```diff
@@ -224,131 +295,93 @@ func (m *monitorT) groupByLatest(policies []model.Policy) map[string]model.Polic
-	// This is by design; there is an optional throttle here.  The queue will roll
-	// out before any new revisions are detected and will slow based on throttle.
-	// Note: We may want a more sophisticated system that detects new revisions during
-	// a throttled rollout; but that is TBD.
-
-	var throttle *time.Ticker
-	if m.throttle != time.Duration(0) {
-		throttle = time.NewTicker(m.throttle)
-		defer throttle.Stop()
-	}
-
-	start := time.Now()
-
-	zlog.Info().
-		Int("nSubs", len(subs)).
-		Dur("throttle", m.throttle).
-		Msg("policy rollout begin")
-
-LOOP:
-	for _, s := range subs {
-
-		if throttle != nil {
-			select {
-			case <-throttle.C:
-			case <-ctx.Done():
-				err = ctx.Err()
-				break LOOP
-			}
-		}
-
-		select {
-		case s.c <- pp:
-		default:
-			// Should never block on a channel; we created a channel of size one.
-			// A block here indicates a logic error somewheres.
-			zlog.Error().
-				Str("policyId", policy.PolicyId).
-				Msg("should never block on policy channel")
-		}
-
+	if newPolicy.CoordinatorIdx <= 0 {
+		zlog.Info().Msg("ignore policy that has not pass through coordinator")
+		return false
 	}
 
-	zlog.Info().
-		Err(err).
-		Dur("tdiff", time.Since(start)).
-		Msg("policy rollout end")
-
-	return err
-}
-
-func (m *monitorT) updatePolicy(pp *ParsedPolicy) []subT {
 	m.mut.Lock()
 	defer m.mut.Unlock()
 
-	newPolicy := pp.Policy
-
 	p, ok := m.policies[newPolicy.PolicyId]
 	if !ok {
 		p = policyT{
 			pp:   *pp,
-			subs: make(map[uint64]subT),
+			head: makeHead(),
 		}
 		m.policies[newPolicy.PolicyId] = p
-		m.log.Info().
-			Str("policyId", newPolicy.PolicyId).
-			Int64("rev", newPolicy.RevisionIdx).
-			Int64("coord", newPolicy.CoordinatorIdx).
-			Msg("new policy")
-		return nil
+		zlog.Info().Msg("new policy added on update")
+		return false
 	}
 
+	// Cache the old stored policy for logging
 	oldPolicy := p.pp.Policy
 
+	// Update the policy in our data structure
 	p.pp = *pp
 	m.policies[newPolicy.PolicyId] = p
 
-	m.log.Info().
-		Str("policyId", newPolicy.PolicyId).
-		Int64("orev", oldPolicy.RevisionIdx).
-		Int64("nrev", newPolicy.RevisionIdx).
-		Int64("ocoord", oldPolicy.CoordinatorIdx).
-		Int64("ncoord", newPolicy.CoordinatorIdx).
-		Msg("policy revised")
+	// Iterate through the subscriptions on this policy;
+	// schedule any subscription for delivery that requires an update.
+	nQueued := 0
 
-	if newPolicy.CoordinatorIdx <= 0 {
-		m.log.Info().
-			Str("policyId", newPolicy.PolicyId).
-			Msg("Do not roll out policy that has not pass through coordinator")
-		return nil
-	}
+	iter := NewIterator(p.head)
+	for sub := iter.Next(); sub != nil; sub = iter.Next() {
+		if sub.isUpdate(&newPolicy) {
+
+			// Unlink the target node from the list
+			iter.Unlink()
+
+			// Push the node onto the pendingQ
+			// HACK: if update is for cloud agent, put on front of queue
+			// not at the end for immediate delivery.
+			if newPolicy.PolicyId == cloudPolicyId {
```

Comment:
Yeah this is an ugly hack. I think you should also add a check to move to the front of the queue `newPolicy.is_default_fleet_server`.

But I think long term we need a way for a priority to be set on the policy in the index. So the cloud policy and the default fleet server policy can set a higher priority than the other policies.